### PR TITLE
Kraus-von Raumer induction for pushouts

### PR DESCRIPTION
**Pull request by ecavallo**
From https://arxiv.org/abs/1901.06022.

The lemma I called "``flippedHomotopyIsInverse``" here has shown up before (``HITs.S1.Base.rotLoopInv`` is an instance) but I don't know if it should be moved up somewhere or what it should be called.

### CODE REVIEW

**Comment by ecavallo** [comment]
This would presumably also be useful for other types like suspension, mapping cone, join. If the lemmas were all moved upstream than the code duplication wouldn't be so bad.

But it doesn't look like transp is being optimized for these special case types, so is there any reason not to define them as instances of the pushout in the first place?

**Comment by m-yac** [comment]
> But it doesn't look like transp is being optimized for these special case types, so is there any reason not to define them as instances of the pushout in the first place?

I noticed this as well, I think it would be nicer for `Susp`, `Cone`, and `Join` to be defined in terms of `Pushout`. We could still include the HIT versions in a comment, like in `Truncation.Base`.

We would probably also want pattern synonyms for each (e.g. `north` for `inl tt`).

File: Cubical/Foundations/Path.agda [hunk 1]
```diff
@@ -17,6 +17,10 @@ cong′ : ∀ {B : Type ℓ'} (f : A → B) {x y : A} (p : x ≡ y)
       → Path B (f x) (f y)
 cong′ f = cong f
 
+PathP→Path : ∀ {ℓ} {A B : Type ℓ} {C : A ≡ B} {a : A} {b : B}
+  → PathP (λ i → C i) a b → transport C a ≡ b
+PathP→Path {C = C} p i = transp (λ j → C (i ∨ j)) i (p i)
```
Comment (mortberg) [inline Cubical/Foundations/Path.agda]: This is kind of the same as: https://github.com/agda/cubical/blob/master/Cubical/Foundations/Prelude.agda#L174

File: Cubical/HITs/Pushout/KrausVonRaumer.agda [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+  (Q : ∀ c → Path (Pushout f g) (inl b₀) (inr c) → Type ℓ''')
+  (r : P b₀ refl)
+  (e : (a : A) (q : inl b₀ ≡ inl (f a)) → P (f a) q ≃ Q (g a) (q ∙ push a))
+  where
+
+  Codes : (d : Pushout f g) (q : inl b₀ ≡ d) → Type ℓ'''
+  Codes (inl b) q = P b q
+  Codes (inr c) q = Q c q
+  Codes (push a i) q =
+    Glue
```
Comment (mortberg) [inline Cubical/HITs/Pushout/KrausVonRaumer.agda]: Can't you use `ua` here?

File: Cubical/HITs/Pushout/KrausVonRaumer.agda [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+  (Q : ∀ c → Path (Pushout f g) (inr c₀) (inr c) → Type ℓ''')
+  (r : Q c₀ refl)
+  (e : (a : A) (q : inr c₀ ≡ inl (f a)) → P (f a) q ≃ Q (g a) (q ∙ push a))
+  where
+
+  Codes : (d : Pushout f g) (q : inr c₀ ≡ d) → Type ℓ'''
+  Codes (inl b) q = P b q
+  Codes (inr c) q = Q c q
+  Codes (push a i) q =
+    Glue
```
Comment (mortberg) [inline Cubical/HITs/Pushout/KrausVonRaumer.agda]: Use `ua`?

File: Cubical/HITs/Pushout/KrausVonRaumer.agda (hunk 2 above)
Comment (ecavallo) [reply Cubical/HITs/Pushout/KrausVonRaumer.agda]: Not as it is, the line being extended is not constant.

File: Cubical/Foundations/Path.agda (hunk 1 above)
Comment (ecavallo) [reply Cubical/Foundations/Path.agda]: Oops, I'll use that one.

**Comment by ecavallo** [comment]
> But why can't we just keep the definitions as they are and have paths between them and the pushout definitions so that we can transport whatever results we want?

Well, any time we do this we negate the efficiency gains of special casing in the first place. And it can be annoying to set up if the theorem talks about the constructors or eliminator.

File: Cubical/HITs/Pushout/KrausVonRaumer.agda [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+
+{-# OPTIONS --cubical --safe #-}
+module Cubical.HITs.Pushout.KrausVonRaumer where
+
+open import Cubical.Foundations.Everything
+open import Cubical.Data.Sigma
+open import Cubical.HITs.Pushout.Base
+
+private
+  flippedHomotopyIsInverse : ∀ {ℓ} {A : Type ℓ} {f : A → A} (p : ∀ a → f a ≡ a)
```
Comment (mortberg) [inline Cubical/HITs/Pushout/KrausVonRaumer.agda:17]: How about moving this to Foundations.Function and calling it "homotopySymInv" or something similar?

**Comment by mortberg** [comment]
> > But why can't we just keep the definitions as they are and have paths between them and the pushout definitions so that we can transport whatever results we want?
> 
> Well, any time we do this we negate the efficiency gains of special casing in the first place. And it can be annoying to set up if the theorem talks about the constructors or eliminator.

Right... So it might be better to use the Pushout definitions after all in order to avoid duplication of proofs. I wouldn't mind keeping the direct definitions as well, but encourage the use of the Pushout ones. No matter what this should be done in future PRs.

File: Cubical/HITs/Pushout/KrausVonRaumer.agda (hunk 4 above)
Comment (ecavallo) [reply Cubical/HITs/Pushout/KrausVonRaumer.agda]: Done! Sadly the other path lemmas are somewhat weird... I hoped that there was a way to build them out of lemmas already in the library but I didn't come up with anything.

**Comment by mortberg** [comment]
Thanks for the contribution!

**Pull request merged**